### PR TITLE
Mvertens/phys cycle

### PR DESCRIPTION
**Pull request by mvertens**
New hooks for runnning external scripts for pre-run and post-run 

CESM POP needs new pre-run and post-run hooks to carry out the lengthy spinup of its ecosystem module. This addition provides additional hooks other than data assimilation to carry this out.
However, it is general enough to be leveraged by other future pre-run and post-run scripts should those be needed.

In addition, all log files for either data assimilation OR the new pre-run and post-run hooks are now output in $RUNDIR rather than $CASEROOT

Test suite: scripts_regression_test AND PRE.f19_f19.ADESP.yellowstone_intel
Test baseline: 
Test namelist changes: 
Test status: bit for bit, roundoff

Fixes: None

User interface changes?: None

Code review:

### CODE REVIEW

**Comment by billsacks** [comment]
@mvertens - thanks a lot for putting this in place. I have a feeling it could be useful for the script used by the LIWG to update CAM's topographic heights before each resubmission. Am I right in understanding the code, that this is still somewhat tied in with the data assimilation capabilities? I'm particularly noticing that the new code is nested under

```python
     for cycle in range(data_assimilation_cycles):
```

which suggests to me that you still need to set data_assimilation_cycles in order to use this new feature... is that true? (I'm not asking you to change that - I'm just trying to understand the usage.)

@JeremyFyke and @lofverstrom - this may be of interest to you, too

**Comment by jedwards4b** [comment]
@billsacks DATA_ASSIMILATION_CYCLES default value is 1, so that loop is always executed at least once.

File: scripts/lib/CIME/case_run.py [hunk 1]
```diff
@@ -208,6 +219,9 @@ def case_run(case):
             case.set_value("CONTINUE_RUN", "TRUE")
             lid = new_lid()
 
+        if prerun_script != "UNSET":
```
Comment (jedwards4b) [inline scripts/lib/CIME/case_run.py]: I've been trying to get rid of UNSET as a default value, preferring None.

File: src/components/data_comps/datm/cime_config/buildnml [hunk 2]
```diff
@@ -147,11 +147,11 @@ def _create_namelists(case, confdir, inst_string, infile, nmlgen):
 
     # set variables that are not per-stream
     if datm_mode == 'CPLHISTForcingForOcnIce':
-        nmlgen.add_default("domainfile", value='null')
         if atm_domain_file != "UNSET":
             if case.get_value('ATM_DOMAIN_FILE') != 'UNSET':
-                case.add_default('ATM_DOMAIN_FILE', value='UNSET', ignore_abs_path=True)
-                case.flush()
+                nmlgen.add_default("domainfile", value=atm_domain_file, ignore_abs_path=True)
```
Comment (jedwards4b) [inline src/components/data_comps/datm/cime_config/buildnml]: This is an unrelated change?

**Comment by mvertens** [comment]
The datm is a bug fix for the phys_cycle spinup work. I never put in an
issue since I just discovered it as part of trying to get the spinup
working out of the box.

**Comment by mvertens** [comment]
@jedwards - are you suggesting that the prerun_script and postrun_script
should be set to None in the xml file as the default value?

**Comment by mvertens** [comment]
@sacks - This is not tied to the data assimilation capabilities and I very
much wanted to keep it separate. As @jedwards pointed out - in
DATA_ASSIMILATION_CYCLES by default is 1- and I envisioned this as either
using data assimilation or the postrun or prerun script.

**Comment by billsacks** [comment]
@jedwards4b and @mvertens - sounds great, thanks for the clarification!

**Comment by jedwards4b** [comment]
Setting it to empty
<default_value/> will result in a None value in python

**Comment by mvertens** [comment]
I have changed the <default_value> to empty - but it does NOT result in a None value in python - it results in a False value. I have verified that the behavior is working with the new empty default setting.
